kdc/fast.c

```diff
@@ -358,133 +446,141 @@ fast_unwrap_request(astgs_request_t r)
-	goto out;
-    if (len != pa->padata_value.length) {
-	ret = KRB5KDC_ERR_PREAUTH_FAILED;
+    if (ret) {
+	kdc_log(r->context, r->config, 4,
+		"Failed to decode PA-FX-FAST-REQUEST: %d", ret);
 	goto out;
     }
 
     if (fxreq.element != choice_PA_FX_FAST_REQUEST_armored_data) {
-	kdc_log(r->context, r->config, 2,
-		"AS-REQ FAST contain unknown type: %d", (int)fxreq.element);
+	kdc_log(r->context, r->config, 4,
+		"PA-FX-FAST-REQUEST contains unknown type: %d",
+		(int)fxreq.element);
 	ret = KRB5KDC_ERR_PREAUTH_FAILED;
 	goto out;
     }
 
-    /* pull out armor key */
-    if (fxreq.u.armored_data.armor == NULL) {
-	kdc_log(r->context, r->config, 2,
+    /*
+     * If check for armor data or it's not a TGS-REQ with implicit
+     * armor.
+     */
+    if (fxreq.u.armored_data.armor == NULL && tgs_ac == NULL) {
+	kdc_log(r->context, r->config, 4,
 		"AS-REQ armor missing");
 	ret = KRB5KDC_ERR_PREAUTH_FAILED;
 	goto out;
     }
 
-    if (fxreq.u.armored_data.armor->armor_type != 1) {
-	kdc_log(r->context, r->config, 2,
-		"AS-REQ armor type not ap-req");
-	ret = KRB5KDC_ERR_PREAUTH_FAILED;
-	goto out;
-    }
-	    
-    ret = krb5_decode_ap_req(r->context,
-			     &fxreq.u.armored_data.armor->armor_value,
-			     &ap_req);
-    if(ret) {
-	kdc_log(r->context, r->config, 2, "AP-REQ decode failed");
-	goto out;
-    }
+    /*
+     *
+     */
+    if (tgs_ac == NULL) {
+	if (fxreq.u.armored_data.armor->armor_type != 1) {
+	    kdc_log(r->context, r->config, 4,
+		    "Incorrect AS-REQ armor type");
+	    ret = KRB5KDC_ERR_PREAUTH_FAILED;
+	    goto out;
+	}
 
-    /* Save that principal that was in the request */
-    ret = _krb5_principalname2krb5_principal(r->context,
-					     &armor_server,
-					     ap_req.ticket.sname,
-					     ap_req.ticket.realm);
-    if (ret) {
-	free_AP_REQ(&ap_req);
-	goto out;
-    }
+	ret = krb5_decode_ap_req(r->context,
+				 &fxreq.u.armored_data.armor->armor_value,
+				 &ap_req);
+	if(ret) {
+	    kdc_log(r->context, r->config, 4, "Failed to decode AP-REQ");
+	    goto out;
+	}
 
-    ret = _kdc_db_fetch(r->context, r->config, armor_server,
-			HDB_F_GET_KRBTGT
-			| HDB_F_DELAY_NEW_KEYS,
-			NULL, NULL, &armor_user);
-    if(ret == HDB_ERR_NOT_FOUND_HERE) {
-	kdc_log(r->context, r->config, 5,
-		"armor key does not have secrets at this KDC, "
-		"need to proxy");
-	free_AP_REQ(&ap_req);
-	goto out;
-    } else if (ret) {
-	free_AP_REQ(&ap_req);
-	ret = KRB5KDC_ERR_S_PRINCIPAL_UNKNOWN;
-	goto out;
-    }
+	/* Save that principal that was in the request */
+	ret = _krb5_principalname2krb5_principal(r->context,
+						 &armor_server,
+						 ap_req.ticket.sname,
+						 ap_req.ticket.realm);
+	if (ret) {
+	    free_AP_REQ(&ap_req);
+	    goto out;
+	}
 
-    ret = hdb_enctype2key(r->context, &armor_user->entry, NULL,
-			  ap_req.ticket.enc_part.etype,
-			  &armor_key);
-    if (ret) {
+	ret = _kdc_db_fetch(r->context, r->config, armor_server,
+			    HDB_F_GET_SERVER, NULL, NULL, &armor_user);
```

Comment:
the armor_server lookup changed from HDB_F_GET_KRBTGT | HDB_F_DELAY_NEW_KEYS to HDB_F_GET_SERVER, was that delibrate.  What principals are permissible here?  (Oddly, this appears to work regardless, given this passes Samba's tests and Samba is the only user of those flags).